File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.java

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.Stack;

public class Main {

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		int n = sc.nextInt();

		Stack<Integer> stack = new Stack<>();

		List<Integer> tmp = new ArrayList<>();

		for (int i = 0; i < n; i++) {
			tmp.add(sc.nextInt());
		}

		for (int i = tmp.size() - 1; i >= 0; i--) {
			stack.push(tmp.get(i));
		}

//		System.out.println(stack);
//		System.out.println(stack.peek());

		Stack<Integer> wait = new Stack<>();
		String ans = "Sad";
		int cnt = 1;
		while (!stack.isEmpty()) {
			if (stack.peek() != cnt) {
				wait.add(stack.pop());
			} else if (stack.peek() == cnt) {
				stack.pop();
				cnt++;
			}

			while (!wait.isEmpty() && wait.peek() == cnt) {
				wait.pop();
				cnt++;
			}

		}
		while (!wait.isEmpty()) {
			if (!wait.isEmpty() && wait.peek() == cnt) {
				wait.pop();
				cnt++;
			}
			else {
				break;
			}
		}

		if (wait.isEmpty()) {
			ans = "Nice";
		}

		System.out.println(ans);
	}

}

File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.py
'''
문제
    큐 1개, stack 1개 를 사용해서 순차적으로 학생들을 빼내라
구상
    1. 줄 서는 곳 q
    2. 대기줄 q
    3. num (1부터시작)
    1,2의 0번째가 num이 되면 빼내기 가능
    1 -> 2 가능 2 -> 1 불가능
'''

n = int(input())
arr = list(map(int, input().split()))
wait = []

num = 1

# num이 n이 되면 다 넘어온 것
ans = "Sad"
stack = []
while num<=n:
    if(arr and arr[0]==num):
        arr.pop(0)
        num+=1
    elif(stack and stack[-1]==num):
        stack.pop()
        num+=1
    else:
        if arr:
            stack.append(arr.pop(0))
        else:
            break
if(not stack and not arr):
    ans = "Nice"
print(ans)
